### PR TITLE
Checked Array Access

### PR DESCRIPTION
**Pull request by mfikes**

### CODE REVIEW

**Comment by swannodette** [comment]
I think the post should be expanded and have a more conversational / philosophical tone. For example we should start with a point about implementation details - like `js*`. When ClojureScript first shipped it was all over the standard library and we slowly backed it out into macros. It's an implementation detail and not something to rely in libraries or application code. The way `aget` appeared in ClojureScript is remarkably similar - kind of a sloppy expedient. Now that we understand the utility and performance characteristic of `goog.object/get` there's just no need to have `get` in the standard library anymore, and we've created a new internal helper `unsafe-get` for the few performance critical cases. We should also allude that by eliminating bad `aget` `aset` usage we can safely add runtime checks during dev as well as align with Clojure behavior with respect to non-integer values.

Basically I think we should tell a story here to give the change some rationale and future purpose.

File: content/news/2017-07-xx-new-aget-aset-compiler-warnings.adoc [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+:jbake-type: post
+
+ifdef::env-github,env-browser[:outfilesuffix: .adoc]
+
+Much of the history of the ClojureScript compiler can be characterized by the 
+themes of pragmatic expediency, followed by successive refinement. This is 
+nicely illustrated by the history of `aget`.
+
+The first minimally viable implementation of `aget` was a simple function. If 
+implemented today, would look like:
```
Comment (sherbondy) [inline content/news/2017-07-xx-new-aget-aset-compiler-warnings.adoc]: > If implemented today, **it** would look like:

(P.S. Love how these draft blog posts are being handled: public & w/ pull-requests, fun sneak preview!)

File: content/news/2017-07-xx-new-aget-aset-compiler-warnings.adoc [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+The ClojureScript standard library itself has places where `aget` and `aset` 
+were being misused for object access, and these are being cleaned up. It turns 
+out that `goog.object/get` is sufficiently performant to replace nearly all 
+uses of `aget` for object access. In the few places where it is not (in highly 
+performance-critical areas in the standard library implmentation), the 
+compiler makes use of a new internal `unsafe-get` macro to get the job done.
+
+We'd encourage you to revise code you control to to ensure that the `aget` and 
+`aset` are used only for array access, and to consider using the facilities in 
+`goog.object` (either directly, or indirectly via a library such as `cljs-
```
Comment (sherbondy) [inline content/news/2017-07-xx-new-aget-aset-compiler-warnings.adoc]: `cljs-oops`

has an unexpected line break in the name here.
might be cool to link to the library too!

Comment (mfikes) [inline content/news/2017-07-xx-new-aget-aset-compiler-warnings.adoc]: Yep. I've combed through the article, hopefully fixing all the c/e errors.

File: content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+The first minimum viable implementation of `aget` was a simple function. If 
+implemented today, it would look like: 
+
+```
+(defn aget [array i]
+  (js* "~{array}[~{i}]"))
+```  
+
+This employs the `js*` special form to directly emit JavaScript which employs
+subscript notation for element access. It looks roughly like this:
```
Comment (swannodette) [inline content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc]: double usage of "employ"

Comment (mfikes) [inline content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc]: Squashing a change that changes the second to "uses".

File: content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+structures by using variadic arguments. Again, the theme of successive refinement.
+
+At some point, a surprising thing started to happen: It was discovered that 
+the implementation of `aget` could be used to access _object_ properties by 
+name. But `aget` was never designed to support this particular use: The 
+“`a`-” family of functions (including `aclone`, `amap`, `areduce`) are all meant 
+for _arrays_, not _objects_. The additional arguments to `aget` are numeric 
+array indices, not string property names. Nevertheless, perhaps owing to the 
+succinctness of forms like
+
```
Comment (swannodette) [inline content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc]: I would probably just start new sentences where you have a ":" here.

Comment (mfikes) [inline content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc]: Fixed

File: content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+introduction of `THROW` and `CATCH` primitives to MacLisp in 1972. The authors 
+go on to say that “the pattern of design (careful or otherwise), unintended 
+use, and later redesign is common.”
+
+What _should_ you use for object property access, if `aget` and `aset` are reserved 
+for arrays? ClojureScript makes the Google Closure library readily accessible, and 
+there are some nice facilities worth checking out in the `goog.object` namespace. In 
+particular, `goog.object/get` and `goog.object/set` are appropriate APIs,
+suited for this purpose. For example, this does what you'd want:
+
```
Comment (swannodette) [inline content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc]: Let's link to the Closure API https://google.github.io/closure-library/api/goog.object.html#get

File: content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+We’d encourage you to revise code you control to to ensure that the `aget` and 
+`aset` are used only for array access, and to consider using the facilities in 
+`goog.object` (either directly, or indirectly via a library such as https://github.com/binaryage/cljs-oops[cljs-oops]) to access object properties.
+
+Towards that end, the upcoming ClojureScript release includes a new optional 
+`:invalid-array-access` warning that indicates,
+when known via type inference, that `aget` or `aset` is
+operating on non-arrays or being supplied non-numeric indices.
+
```
Comment (swannodette) [inline content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc]: We might want to add - "in the future setting this warning may also cause emissions of checked array access during development as we suggested above".

File: content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+nicely illustrated by the history of `aget`.
+
+The first minimum viable implementation of `aget` was a simple function. If 
+implemented today, it would look like: 
+
+```
+(defn aget [array i]
+  (js* "~{array}[~{i}]"))
+```  
+
```
Comment (swannodette) [inline content/news/2017-07-xx-proper-array-and-object-access-patterns.adoc]: need `[source,clojure]` on this line and the same for the snippets but with the correct language setting.

**Comment by mfikes** [comment]
Currently titled "Proper Array and Object Access Patterns"
With the updates now currently in https://dev.clojure.org/jira/browse/CLJS-2198, perhaps a better title for this post might be along the lines of "Enhanced Array Access Validation"

**Comment by swannodette** [comment]
I like "Checked Array Access", short, sweet and matches the compiler option?

File: content/news/2017-07-14-checked-array-access.adoc [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+This is the third post in the 
+https://clojurescript.org/news/2017-07-07-sneak-preview[Sneak Preview] series.
+
+Much of the history of the ClojureScript compiler can be characterized by the 
+themes of pragmatic expediency, followed by successive refinement. This is 
+nicely illustrated by the history of `aget`.
+
+The first minimum viable implementation of `aget` was a simple function. If 
+implemented today, it would look like: 
+
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: "If implemented, today it would like:" is strangely worded I think - this *was* how it was implemented, see https://github.com/clojure/clojurescript/blob/86ad57839aa471b03d17c7ce1418736c7e9c7388/src/cljs/cljs/core.cljs#L603-L604.

I would change this to say:

Six years ago it looked

```clojure
(defn aget [array i]
   (js* "return ~{array}[~{i}]"))
```

File: content/news/2017-07-14-checked-array-access.adoc [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+[source,clojure]
+```
+(defn aget [array i]
+  (js* "~{array}[~{i}]"))
+```  
+
+This employs the `js*` special form to directly emit JavaScript which uses
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: employs the *internal* `js*` special form

File: content/news/2017-07-14-checked-array-access.adoc [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+subscript notation for element access. It looks roughly like this:
+
+[source,javascript]
+```
+function aget(array, i) {
+  return array[i];
+}
+```
+
+> Note that `js*` is not intended to be used in application-level ClojureScript code—it is an internal mechanism used by the compiler.
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: drop after the &mdash;

File: content/news/2017-07-14-checked-array-access.adoc [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+Subsequently, the `js*` form was further enhanced to support expression 
+embedding, and the very first use of this new `js*` capability was in the 
+introduction of the macro alternative of `aget`:
+
+[source,clojure]
+```
+(defmacro aget [a i]
+  (list 'js* "(~{}[~{}])" a i))
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: I think this part about the macro variant is inessential let's just drop it.

File: content/news/2017-07-14-checked-array-access.adoc [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+Interestingly, in the early history of the compiler, `js*` was employed fairly
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: Drop "interestingly"

File: content/news/2017-07-14-checked-array-access.adoc [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+[source,clojure]
+```
+(defmacro aget [a i]
+  (list 'js* "(~{}[~{}])" a i))
+```
+
+Interestingly, in the early history of the compiler, `js*` was employed fairly 
+heavily in the runtime functions that ship with the ClojureScript standard 
+library. Over time, many uses of `js*` were backed out into macros as 
+the implementation was tuned to emit highly efficient JavaScript.
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: Over time, raw usage of `js*` was removed entirely in the standard library and hidden behind reusable macros.

File: content/news/2017-07-14-checked-array-access.adoc [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+As time moved on, our friend `aget` (as well as `aset`) was made to more 
+closely match Clojure by allowing it to access nested array
+structures by using variadic arguments. Again, the theme of successive refinement.
+
+At some point, a surprising thing started to happen: It was discovered that
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: Readers familiar with JavaScript will note that the above `aget` implementation works perfectly well for JavaScript objects. This fact, like `js*` was also abused in the early days of the standard library. Unfortunately due to insufficient documentation about alternative, users began to mimic this internal detail.

File: content/news/2017-07-14-checked-array-access.adoc [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+As time moved on, our friend `aget` (as well as `aset`) was made to more 
+closely match Clojure by allowing it to access nested array
+structures by using variadic arguments. Again, the theme of successive refinement.
+
+At some point, a surprising thing started to happen: It was discovered that 
+the implementation of `aget` could be used to access _object_ properties by 
+name. But `aget` was never designed to support this particular use. The 
+“`a`-” family of functions (including `aclone`, `amap`, `areduce`) are all meant 
+for _arrays_, not _objects_. The additional arguments to `aget` are numeric 
+array indices, not string property names. Nevertheless, perhaps owing to the
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: perhaps to the lure of ease, forms like

File: content/news/2017-07-14-checked-array-access.adoc [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+for _arrays_, not _objects_. The additional arguments to `aget` are numeric 
+array indices, not string property names. Nevertheless, perhaps owing to the 
+succinctness of forms like
+
+[source,clojure]
+```
+(aget #js {:foo 1} "foo")
+```
+
+`aget` became heavily used in the wild to access JavaScript data, or to avoid the
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: `aget` became heavily used in the wild to avoid

File: content/news/2017-07-14-checked-array-access.adoc [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+heavily in the runtime functions that ship with the ClojureScript standard 
+library. Over time, raw usage of `js*` was removed entirely in the standard 
+library and hidden behind reusable macros.
+
+As time moved on, our friend `aget` (as well as `aset`) was made to more 
+closely match Clojure by allowing it to access nested array
+structures by using variadic arguments. Again, the theme of successive refinement.
+
+Readers familiar with JavaScript will note that the above aget implementation
+works perfectly well for JavaScript objects. This fact, like `js*` was also
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: This fact, like `js*`,

File: content/news/2017-07-14-checked-array-access.adoc [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+abused in the early days of the standard library. Unfortunately due to
```
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: Unfortunately,

File: content/news/2017-07-14-checked-array-access.adoc [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+library and hidden behind reusable macros.
+
+As time moved on, our friend `aget` (as well as `aset`) was made to more 
+closely match Clojure by allowing it to access nested array
+structures by using variadic arguments. Again, the theme of successive refinement.
+
+Readers familiar with JavaScript will note that the above aget implementation
+works perfectly well for JavaScript objects. This fact, like `js*`, was also
+abused in the early days of the standard library. Unfortunately, due to 
+insufficient documentation about alternative, users began to mimic this
```
Comment (mfikes) [inline content/news/2017-07-14-checked-array-access.adoc]: about alternatives

Needs plural?

File: content/news/2017-07-14-checked-array-access.adoc [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+> Note that `js*` is not intended to be used in application-level ClojureScript code.
+
+In the early history of the compiler, `js*` was employed fairly 
+heavily in the runtime functions that ship with the ClojureScript standard 
+library. Over time, raw usage of `js*` was removed entirely in the standard 
+library and hidden behind reusable macros.
+
+As time moved on, our friend `aget` (as well as `aset`) was made to more 
+closely match Clojure by allowing it to access nested array
+structures by using variadic arguments. Again, the theme of successive refinement.
```
Comment (mfikes) [inline content/news/2017-07-14-checked-array-access.adoc]: "Again" here seems odd

File: content/news/2017-07-14-checked-array-access.adoc (hunk 19 above)
Comment (swannodette) [inline content/news/2017-07-14-checked-array-access.adoc]: Yes
